keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+    ```
+  - *Cons*:
+    - No multi-arch support, builds only `linux/amd64`
+    - rpm packaging needs `rpmbuild` to be installed on the host
+  - *Pros*:
+    - `bazel` integration
+    - picks up proper version from git
+    - package specs and additional files (systemd units, ...) are versioned in 
+  - *Parts to reuse*:
+    - some sort of `bazel` integration
+    - versioning of the packaging specs and additional files alongside the code
+
+[k-release]: https://github.com/kubernetes/release/tree/master/
+[k-k]: https://github.com/kubernetes/kubernetes/tree/master/build/
+[k-release-deb]: https://github.com/kubernetes/release/tree/master/debian
+[k-release-rpm]: https://github.com/kubernetes/release/tree/master/rpm
+[k-k-deb]: https://github.com/kubernetes/kubernetes/tree/master/build/debs
+[k-k-rpm]: https://github.com/kubernetes/kubernetes/tree/master/build/rpms
+
+To a good part this KEP tries to address most the [issue "k/k is the canonical source of all build artefacts"](https://github.com/kubernetes/kubernetes/issues/71677) and related
+issues:
+- [`kubeadm` package should depend on the same version of `kubelet` and
+  `kubectl` packages](https://github.com/kubernetes/kubernetes/issues/72871)
+- [Automate deb/rpm publishing](https://github.com/kubernetes/sig-release/issues/10)
+- ...
+
+
+
+### Goals
+
+- Fold the packaging specs, systemd unit files, dependency configuration, and
+  "Parts to reuse" into one place in `k/k`.
+
+  As a first step, the state from [`k/release/{rpm,debian}`][k-release] should
+  be used, as those are the things we release right now. As a second step all
+  the changes and additions from [`k/k/build/`][k-k] should be merged in.
+
+  There must be **one** package configuration per package, all different
+  packaging formats must use this as source of truth (a package configuration
+  could be *as an example* a config file for a package-generator plus
+  all the needed systemd unit files, ... )
+
+- The tooling must be used for cutting packages in the official release
+  process, it must also be usable in local development to generate one-off
+  packages.
+
+- <a id="pkg-list"></a>The tooling must support packaging
+  - `kubelet`
+  - `kubectl`
+  - `kubeadm`
+  - `cri-tools`
+  - `kubernetes-cni`
+
+  It should be easy to integrate new packages, especially when they are
+  one-binary packages
+
+- <a id="plat-list"></a>The tooling must support packaging for multiple platforms
+  - `linux/amd64`
+  - `linux/arm64`
+  - `linux/armhf`
+  - `linux/ppc64el`
+  - `linux/s390x`
+
+  It should be easy to integrate other platforms, given there is general
+  support by the supported distributions and the cross-building tools.
+  Examples: `kubectl` for `darwin/amd64`, `kubelet` for `windows/386`, ...
+
+- The tooling must support building `deb`s and `rpm`s
+
+  It should be possible to add new package types, e.g. NuGet for Windows
+
+- The tooling must support signing the generated packages
```

Comment:
signing can be a separate step imo not a strict requirement.

Reply:
You are right.

I was thinking that maybe for signing the package (in contrary to the repository metadata) should be done at the time we actually create the package, however i think:
- we don't necessarily want to do that (at least in the case of the debs, as the verification of those signatures seems to be off by default anyway)
- we always can do it at a later time, e.g. just before publishing

Reply:
As far as I can tell from talking to people and some "research":
- deb don't need to be signed at package build time, the default way is to sign the repo metadata on publishing time
- rpm can also be signed after being created, with [`rpm --addsign`][rpm-addsign], yum repo metadata can/should also be signed

So: The signing step can be done after package creation at the publishing step.

cc: @dims @chris-short

[rpm-addsign]: http://ftp.rpm.org/max-rpm/s1-rpm-pgp-signing-packages.html#S2-RPM-PGP-ADDSIGN-OPTION

Reply:
For debs; Repo metadata must be signed
For RPMs: RPMs must be signed, repo metadata optionally, but best practice to sign.

Makes sense for packaging to happen at publishing time, after it's passed through the various gates. I think in the past, we've talked about votes from approved key holders that then unseals an appropriate signing key from a HSM.